django/utils/deprecation.py

```diff
@@ -83,6 +85,187 @@ def __new__(cls, name, bases, attrs):
+            for name in remappable_names
+        ):
+            raise TypeError(
+                "@deprecate_posargs() remappable_names must"
+                " all be keyword-only parameters."
+            )
+
+        num_remappable_args = len(remappable_names)
+        max_positional_args = num_positional_params + num_remappable_args
+
+        func_name = func.__name__
+        if func_name == "__init__":
+            # In the warning, show "ClassName(...)" rather than "__init__(...)".
+            # The class isn't defined yet, but its name is in __qualname__.
+            # Some examples: "ClassName.__init__", "Nested.ClassName.__init__",
+            # "MyTests.test_case.<locals>.ClassName.__init__".
+            local_name = func.__qualname__.rsplit("<locals>.", 1)[-1]
+            class_name = local_name.replace(".__init__", "")
+            func_name = class_name
+
+        def remap_deprecated_args(args, kwargs):
+            """
+            Move deprecated positional args to kwargs and issue a warning.
+            Returns updated (args, kwargs).
+            """
+            num_positional_args = len(args)
+            if num_positional_args > max_positional_args:
+                raise TypeError(
+                    f"{func_name}() takes"
+                    f" at most {max_positional_args} positional argument(s)"
+                    f" (including {num_remappable_args} deprecated)"
+                    f" but {num_positional_args} were given."
+                )
+
+            # Identify which of the potentially _remappable_ params
+            # are actually being _remapped_ in this particular call.
+            remapped_names = remappable_names[
+                : num_positional_args - num_positional_params
+            ]
+            conflicts = set(remapped_names) & set(kwargs)
+            if conflicts:
+                # Report duplicate param names in original parameter order.
+                conflicts_str = ", ".join(
+                    f"'{name}'" for name in remapped_names if name in conflicts
+                )
+                raise TypeError(
+                    f"{func_name}() got both deprecated positional and keyword"
+                    f" argument values for {conflicts_str}."
+                )
+
+            # Do the remapping.
+            remapped_kwargs = dict(zip(remapped_names, args[num_positional_params:]))
+            remaining_args = args[:num_positional_params]
+            updated_kwargs = kwargs | remapped_kwargs
+
+            # Issue the deprecation warning.
+            message_template = (
+                message_template_singular
+                if len(remapped_names) == 1
+                else message_template_plural
+            )
+            message = message_template.format(
+                remapped_names=", ".join(f"'{name}'" for name in remapped_names),
+                func_name=func_name,
+            )
+            warnings.warn(message, deprecation_warning, stacklevel=3)
```

Comment:
Following ticket-35667 and comments from @charettes like [this one](https://github.com/django/django/pull/19187#discussion_r2175056745), could we change this to use `skip_file_prefixes=(os.path.dirname(django.__file__),` instead of `stacklevel=3`? So:

```suggestion
            warnings.warn(
                message,
                deprecation_warning,
                skip_file_prefixes=(os.path.dirname(django.__file__),),
            )
```

Then, ideally, tests should be adjusted to assert over the deprecation's reported filename, as documented in https://docs.djangoproject.com/en/5.2/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature:

```python
def test_foo_deprecation_warning(self):
    msg = "Expected deprecation message"
    with self.assertWarnsMessage(RemovedInDjangoXXWarning, msg) as ctx:
        # invoke deprecated behavior
        ...
    self.assertEqual(ctx.filename, __file__)
```

Reply:
> `skip_file_prefixes`

Oh, nice, TIL.

The test is already there (currently called test_warning_stacklevel; I'll rename it), and passes both before and after changing stacklevel to skip_file_prefixes.